Fastly configuration

jhadobe opened the pull request:
## Purpose of this pull request

This PR adds instructions for for configuring Fastly as a CDN in Adobe Commerce with API Mesh.

Pages affected: 
- [Edge Caching](https://developer.adobe.com/graphql-mesh-gateway/gateway/cache-control-headers/)

Staging site: https://adobedocs.github.io/graphql-mesh-gateway/gateway/fastly/

jhadobe commented:
@amolina-adobe @cpartica my internal cloud does not have access to Fastly, so I was unable to confirm these instructions.

amolina-adobe commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,251 @@
+
+### Fastly Prefixing
+
+API mesh prefixes any Fastly source headers with their source name. For example, a source named "commerce" with an `x-magento-cache-id` header is converted to `x-commerce-magento-cache-id`. However, if your endpoint URL contains "magento", API Mesh assumes you are connecting to an Adobe Commerce instance and does not prefix your headers with a source name. Using the previous example, your header  ...
```
```suggestion
API Mesh prefixes any Fastly source headers with their source name. For example, a source named "commerce" with an `x-magento-cache-id` header is converted to `x-commerce-magento-cache-id`. However, if your endpoint URL contains "magento", API Mesh assumes you are connecting to an Adobe Commerce instance and does not prefix your headers with a source name. Using the previous example, your header would remain `x-magento-cache-id`.
```

amolina-adobe commented on src/pages/gateway/fastly.md:
````diff
@@ -0,0 +1,251 @@
+}
+```
+
+### Fastly Prefixing
````
Does it makes sense to change this header? Although it may affect some of the caching behavior, the prefixing isn't specific to Fastly. Maybe something like the following?
```suggestion
### Cache Header Prefixing
```
I don't feel particularly strong about this so open to other ideas.

amolina-adobe commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,251 @@
+
+After adding VCL snippets in the [Fastly Setup](#configure-fastly-in-adobe-commerce), your Commerce GraphQL URL is now in the following format with no `api_key` appended: `https://graph.adobe.io/api/<meshId>/graphql`
+
+To distinguish between requests from users and requests from API Mesh, use the following configuration option to prevent Fastly from caching headers that come directly from API Mesh:
```
Perhaps mention that this is a source operationHeader to help the user understand where to add it

amolina-adobe commented on src/pages/gateway/fastly.md:
````diff
@@ -0,0 +1,251 @@
+        # }
+        ```
+
+    The remaining snippets allow you to query your mesh URL without appending the `api_key`:
````
Not a big deal, but I think this is only true for the `api_mesh_pass` snippet. It's essentially a duplicate of the `api_mesh_miss` snippet which routes to the appropriate backend and set the api key when querying a mesh.

jhadobe commented:
thanks @amolina-adobe! I believe I have addressed your comments. I attempted to tidy up the `vcl` snippets where possible. I'll hold off on making changes in response to your comments where we need @cpartica's input.

cpartica commented on src/data/navigation/sections/gateway.js:
```diff
@@ -43,6 +43,10 @@ module.exports = [
                 title: 'Edge caching and cache control',
```
I think we should rename edge caching term entirely to make space for Dynamic content caching, which could be done through CDN content delivery and API Mesh caching  (which we call server-side caching - IDK how to call that - if we do move API Mesh to the cloudflare it will become an edge caching - but not necessarily done through a CDN)

The feature we're describing as edge caching is simply being enabled by cache control headers which can be used by a CDN or client (browser) caching

cpartica commented on src/pages/gateway/cache-control-headers.md:
```diff
@@ -13,11 +13,11 @@ keywords:
 <InlineAlert variant="info" slots="text"/>
 
-When using a CDN, you must invalidate the cache after modifying a mesh configuration or you will receive stale information.
+When using a CDN, you must invalidate the cache after modifying a mesh configuration, or you will receive stale information.
```
is Edge caching = CDN here then or what other types of edge caching did we refer to?
see how this is ambiguous?

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,257 @@
+
+# Fastly configuration
+
+Using [Fastly](https://experienceleague.adobe.com/docs/commerce-cloud-service/user-guide/cdn/fastly.html?lang=en) as a content delivery network (CDN) with API Mesh for Adobe Developer App Builder provides additional security and improved performance. To fully integrate API Mesh, Fastly, and Adobe Commerce, you must follow the configuration instructions in the following sections.
```
a content delivery network (CDN) for caching dynamic content (API)

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,257 @@
+
+## Configure Fastly in Adobe Commerce
+
+Use the following process to configure edge caching for Adobe Commerce using the provided Fastly CDN.
```
edge caching should be here dynamic content - as it's not based on the headers that API mesh outputs like explained in [cache-control-headers.md](https://github.com/AdobeDocs/graphql-mesh-gateway/pull/161/files#diff-c2d1334ee9d83dd922562bc477db7e2f8a72e8c241f1f9be050cb3a7b6ef0f7f)

edit:
following steps process not process

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,257 @@
+
+Use the following process to configure edge caching for Adobe Commerce using the provided Fastly CDN.
+
+1. [Get your Fastly credentials](https://experienceleague.adobe.com/docs/commerce-cloud-service/user-guide/cdn/setup-fastly/fastly-configuration.html?lang=en#get-fastly-credentials).
```
You will need to get your provided credentials (Fastly Service ID** and **Fastly API Token )which usually come included with Adobe Commerce

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,257 @@
+
+1. [Get your Fastly credentials](https://experienceleague.adobe.com/docs/commerce-cloud-service/user-guide/cdn/setup-fastly/fastly-configuration.html?lang=en#get-fastly-credentials).
+
+1. In the Adobe Commerce Admin, select **Store** > Settings > **Configuration** > **Advanced** > **System** > **Full Page Cache** and set the **Caching Application** field to **Fastly CDN**.
```
may be worth to provide a screenshot?

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,251 @@
+
+1. Enter the Fastly credentials you retrieved previously into the **Fastly Service ID** and **Fastly API Token** fields. Then click the Test Credentials button to verify that your credentials are correct.
+
+1. Under **Fastly Configuration** > **Custom VCL Snippets** click **Create** and add the following [snippets](https://experienceleague.adobe.com/docs/commerce-cloud-service/user-guide/cdn/custom-vcl-snippets/fastly-vcl-custom-snippets.html). For more information on VCL subroutines, see [Custom Subroutines](https://developer.fastly.com/reference/vcl/subroutines/).
```
no, because you can't modify that with the UI - it hides the custom VCL that's generated by Fastly module 
SO there's no way to update..
we also want to prevent patching code - which would be the other way to accomplish this- so we will use new Custom VCL Snippets for all modifications and play with the priority to get around existing VCL to overriding existing logic

cpartica commented on src/pages/gateway/fastly.md:
````diff
@@ -0,0 +1,257 @@
+     - **Content**:
+  
+        ```csharp
+        backend F_graph_prod_adobe_io {
````
we should probably put this first in the list since is the most important 
like do some sort of logic order not just provide a bulk code
the idea is to also explain each snippet a bit - what it does which we do a bit in the title:
Allows API Mesh to function as a backend
May be worth linking to Fastly backends docs?

cpartica replied:
@cpartica to add a diagram concept

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,257 @@
+
+After adding VCL snippets in the [Fastly Setup](#configure-fastly-in-adobe-commerce), your Commerce GraphQL URL is now in the following format with no `api_key` appended: `https://graph.adobe.io/api/<meshId>/graphql`
+
+To distinguish between requests from users and requests from API Mesh, use the following source operation header to prevent Fastly from caching headers that come directly from API Mesh:
```
here ..we'll be switching experiences/products
so we can say ..next we will focus on API Mesh configuration..or something like that
Then you can explain why do we need to modify API Mesh like you did here

cpartica commented on src/pages/gateway/fastly.md:
````diff
@@ -0,0 +1,257 @@
+To distinguish between requests from users and requests from API Mesh, use the following source operation header to prevent Fastly from caching headers that come directly from API Mesh:
+
+```json
+"x-commerce-bypass-fastly-cache": "true" 
````
we can explain here that this prevents the double caching problem
it's worth including a diagram..
that helped immediately to anybody that I showed this to - they immediately understood it

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,257 @@
+        if (req.url~"^/api/<mesh_id>") {
+          set bereq.http.x - api - key = "<mesh_api_key>";
+        }
+        # //Optionally add another environment
```
environment really means different api meshes

cpartica commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,257 @@
+
+### Cache Header Prefixing
+
+API Mesh prefixes any Fastly source headers with their source name. For example, a source named "commerce" with an `x-magento-cache-id` header is converted to `x-commerce-magento-cache-id`. However, if your endpoint URL contains "magento", API Mesh assumes you are connecting to an Adobe Commerce instance and does not prefix your headers with a source name. Using the previous example, your header  ...
```
let's include a small test on how to know if it worked or not..

jhadobe commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,297 @@
+
+### Test your mesh
+
+After you [create](./create-mesh.md#create-a-mesh) or [update](./create-mesh.md#update-an-existing-mesh) your mesh, run an [`aio api-mesh:describe`](./command-reference.md#aio-api-meshdescribe) command to view your mesh URL. Run a query against this URL to confirm that your mesh is working correctly.
```
Should we provide a specific query?

cpartica replied:
yeah - we can point to a product query

jhadobe commented:
@cpartica @amolina-adobe I have updated this PR with the changes we discussed earlier, with two outstanding exceptions:
- Adding a diagram
- Verify the appropriate cURL timer

keharper commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,297 @@
+   - Cache miss - replace the `<mesh_id>` and `<mesh_api_key>` placeholders with the information from your API Mesh URL, which has the following structure: `https://graph.adobe.io/api/<meshId>/graphql?api_key=<your_apiKey>`. You can retrieve this information by running an [`aio api-mesh:describe`](./command-reference.md#aio-api-meshdescribe) command:
+     - **Name** - api_mesh_miss
+     - **Type** - **miss**
+     - **Priority** - **60**
```
You have multiple snippets with a priority value of 60 and 10 (maybe others). Is that OK?

jhadobe replied:
I believe the priority value is specific to the Fastly operator. @amolina-adobe can you confirm?

amolina-adobe replied:
I think that's fine. The values provided are for a working configuration. For reference, all snippets added through the Adobe Commerce admin get a default priority of 50 unless otherwise specified. https://experienceleague.adobe.com/docs/commerce-cloud-service/user-guide/cdn/custom-vcl-snippets/fastly-vcl-custom-snippets.html?lang=en
![image](https://github.com/AdobeDocs/graphql-mesh-gateway/assets/44038475/5f2a503e-b5fd-4e90-b862-5ff9c064f4e8)

keharper commented on src/pages/gateway/fastly.md:
````diff
@@ -0,0 +1,297 @@
+
+Run the following cURL command, replacing `<Commerce-URL>` with your Adobe Commerce storefront URL.
+
+```typescript
````
```suggestion
```bash
```

jhadobe commented on src/pages/gateway/fastly.md:
```diff
@@ -0,0 +1,297 @@
+
+<InlineAlert variant="info" slots="text"/>
+
+Use the `fastly-debug:1` request header to get more information from Fastly on each request.
```
@amolina-adobe do you think we should add guidance around not using debug headers in production meshes?

amolina-adobe replied:
I don't know the answer to this. This was a value that @cpartica included when configuring. My intuition says it probably shouldn't be used for production.

jhadobe commented on src/pages/gateway/fastly.md:
````diff
@@ -0,0 +1,297 @@
+To distinguish between requests from users and requests from API Mesh, use the following source operation header to prevent Fastly from caching headers that come directly from API Mesh:
+
+```json
+"x-commerce-bypass-fastly-cache": "true" 
````
no, we don't. good idea!
